pkg/cloud/cloudcheck/cloudcheck_processor.go

```diff
@@ -52,24 +65,49 @@ func checkURI(
 	opener cloud.ExternalStorageFromURIFactory,
 	location string,
 	username username.SQLUsername,
-	transferSize int64,
+	params TestParams,
 ) result {
+	ctxDone := ctx.Done()
+
+	transferSize := params.TransferSize
 	if transferSize == 0 {
 		transferSize = 32 << 20
 	}
 
-	var res result
-	store, err := opener(ctx, location, username)
-	if err != nil {
-		res.error = errors.Wrapf(err, "opening external storage").Error()
-		return res
-	}
-	defer store.Close()
-	res, err = checkStorage(ctx, store, transferSize)
-	if err != nil {
-		res.error = err.Error()
+	var total result
+
+	start := timeutil.Now()
+	for {
+		select {
+		case <-ctxDone:
+			return total
+		default:
+		}
+
+		store, err := opener(ctx, location, username)
+		if err != nil {
+			total.error = errors.Wrapf(err, "opening external storage").Error()
+			return total
+		}
+		defer store.Close()
+
+		res, err := checkStorage(ctx, store, transferSize)
+		if err != nil {
+			res.error = err.Error()
+		}
+
+		if total.ok {
```

Comment:
If it's not OK, shouldn't we return early? Or at least store the error message? Overwriting seems not quite right

Reply:
we do return early on the next line when ok is false after we set it here, so we're never overwriting anything other than a zero-value here. restructured it to be more explicit.